vm/ieleApi.ml

```diff
@@ -0,0 +1,157 @@
+open IeleClientUtils
+open Yojson.Basic.Util
+open Msg_types
+
+type block = {
+  state: (string * Yojson.Basic.json) list;
+  timestamp: string
+}
+
+let blocks = ref [{state=[];timestamp=""}]
+
+let nextTimestamp = ref None
+let nextBeneficiary = ref "0x0000000000000000000000000000000000000000"
+
+let pendingTx = ref []
+
+module StringMap = Map.Make(String)
+
+let receipts = ref StringMap.empty
+
+(* not secure, but we should only use this for testing purposes *)
+let random = Cryptokit.Random.pseudo_rng "1234567890123456"
+
+let get_block blocknumber = 
+  let num_int = if blocknumber = "latest" then 0 else (List.length !blocks) - (int_of_string blocknumber) - 1 in
+  List.nth !blocks num_int
+
+let personal_newAccount () =
+  let acct = Cryptokit.Random.string random 20 in
+  let key = to_hex (Bytes.of_string acct) in
+  let value = `Assoc [("nonce", `String "0x00"); ("balance", `String "0x00"); ("storage", `Assoc []); ("code", `String "")] in
+  let latest = List.hd !blocks in
+  let state = latest.state in
+  let new_state = (key, value) :: state in
+  blocks := {state=new_state;timestamp=latest.timestamp} :: List.tl !blocks;
+  `String (to_hex (Bytes.of_string acct))
+
+let setBalance (address, account) =
+  let balance = account |> member "wei" |> to_string in
+  let latest = List.hd !blocks in
+  let state = latest.state in
+  let value = List.assoc address state in
+  let list = value |> to_assoc in
+  let list_without = List.remove_assoc "balance" list in
+  let new_account = `Assoc (sort_assoc_list (("balance", `String balance) :: list_without)) in
```

Comment:
Why is this sorted, when the list of accounts isn't?

Reply:
I believe because I thought that the code that reads from this json depends on the order of fields, but I forget. I can try removing it and see if it breaks anything if you want?